src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace Symfony\Component\AutoMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\AutoMapper\Exception\RuntimeException;
+use Symfony\Component\AutoMapper\Attributes\MapIf;
+use Symfony\Component\AutoMapper\Attributes\MapTo;
+use Symfony\Component\AutoMapper\Attributes\MapWith;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @experimental
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class AutoMapper implements AutoMapperInterface
+{
+    public function __construct(private ?PropertyAccessorInterface $propertyAccessor = null, private ?ContainerInterface $serviceLocator = null)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function map(object $object, object|string|null $to = null): mixed
+    {
+        $refl = new \ReflectionClass($object);
+
+        if (!$to) {
+            $to = $this->getAttribute($refl, MapTo::class, true)->to;
+        }
+
+        $arguments = [];
+        if (is_object($to)) {
+            $toRefl = new \ReflectionClass(get_class($to));
+            $mapped = $to;
+        } else {
+            $toRefl = new \ReflectionClass($to);
+            $mapped = $toRefl->newInstanceWithoutConstructor();
```

Comment:
We need a hook here to be able to "load" the object via some callable. The use case is:

```php
class ProductDto
{
    public CategoryDto $category;
}
```

You're mapping back to a `Product` entity and your `ProductDto.category` property is a `CategoryDto` with `id=5` representing a pre-existing category. So, when we get to mapping `ProductDto.category`, we need a way to see that `CategoryDto` has an id, and query the database for it instead of instantiating a new one. 

We could maybe have `#[MapWith]` above the `category` property that calls a service that queries for the `Category` object. But you'd need to do this above every property like this... and you would also then need to inject the mapper into that so you could continue mapping that object:

```
$category = $this->categoryRepository($categoryDto->id);

$this->mapper->map($categoryDto, $category);
```

Maybe... that's ok? Maybe there is a built in `MapTo(EntityMapper::class)` service that could do this for you 🤔

Reply:
It works already as when calling `map` on underlying objects we use that object instead of calling this or I missed something, I need to try that in a real use case to confirm.

Reply:
When, in this example, that `CategoryDto` is mapped, the final code is effectively:

```php
$mapper->map($categoryDto, Category::class)
```

So that will cause a `new Category` to be created with no hook point (I think) for me to say "Hey! I need to actually load that `Category` object from the database / some other source".

Ref: https://github.com/symfony/symfony/pull/51741/files#diff-32729f6c6e16ff35ea82d8cf0ee4b24313f6c6f4e4ec6b833ec3592880c756bcR70-R71

Reply:
I'd personally completely remove the `object` type hint, and let the `map` function always behave as a factory, in order to keep things simple.

Idea being that your objects might carry readonly properties, and having the object to populate passed the map method will not work in this case.

This would prevent the "load from repository then update object from DTO" use case, I don't see any automatic non magical way of doing this anyway, any suggestions ?

Reply:
Not sure to understand why it wouldn't work I'd the props are writable from the constructor can you give me code corresponding to the use case you've in mind ?

Reply:
I get that the `map()` function with an object to be hydrated as a second parameter means that the mapper will act as an hydrator; I didn't read the whole code so I might be wrong here, but in the case you'd attempt to a set a property using some scope stealing closure magic or by a direct public property access, it'd fail since it wouldn't use the constructor (it cannot since the object has already been created, if I guessed right) ?

Reply:
I think that this is solved by my last update, you can even do a transform if you need more control